server/src/main/java/org/elasticsearch/transport/ProxyConnectionStrategy.java

```diff
@@ -252,15 +258,27 @@ public void onResponse(Void v) {
                         if (shouldOpenMoreConnections()) {
                             openConnections(finished, attemptNumber + 1);
                         } else {
+                            assert connectionManager.size() > 0 : "must have at least one opened connection";
                             finished.onResponse(v);
                         }
                     }
                 }
 
                 @Override
                 public void onFailure(Exception e) {
+                    exceptions.put(new Tuple<>(e.getClass(), e.getMessage()), e);
                     if (countDown.countDown()) {
-                        openConnections(finished, attemptNumber + 1);
+                        if (attemptNumber >= MAX_CONNECT_ATTEMPTS_PER_RUN && connectionManager.size() == 0) {
+                            logger.warn(() -> "failed to open any proxy connections to cluster [" + clusterAlias + "]", e);
```

Comment:
Sanity checking that `warn` is the right log level: this could be caused by longer-lasting but still temporary network glitches, so I'm wondering if we want `info` or even `debug` here instead.

Reply:
`warn` is good. `info` shouldn't really be used to report exceptions, and `debug` is hidden by default so will make troubleshooting too difficult. See [these docs](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#info) for more info.

Reply:
cheers for the pointers, all good to keep this at `warn` here!